File: ios/Packages/MemoryMap/Sources/DataPhoto/CoversFile.swift
import CoreModel
import Foundation

/// `covers.json` 의 모양. `{"region_11140":"a1b2c3", "day_2026-03-05":"d4e5f6"}`
///
/// 서버에 두든 기기에 두든 **같은 형식**입니다. 혼자 쓰던 짜국을 나중에 둘이로 바꿀 때
/// 파일을 그대로 올리면 되도록 하려는 것입니다 (`docs/app/AUTH.md`).
/// 안드로이드 `CoversFile` 과 같은 규칙입니다.
enum CoversFile {

    static func parse(_ data: Data) -> [Cover] {
        guard let raw = try? JSONSerialization.jsonObject(with: data) as? [String: String] else {
            return []
        }
        return raw.compactMap { documentId, photoId in
            guard let key = key(from: documentId) else { return nil }
            return Cover(key: key, photoId: PhotoId(photoId))
        }
    }

    static func data(_ covers: [Cover]) -> Data? {
        let dictionary = Dictionary(covers.map { ($0.key.documentId, $0.photoId.value) },
                                    uniquingKeysWith: { _, last in last })
        return try? JSONSerialization.data(withJSONObject: dictionary)
    }

    static func key(from documentId: String) -> CoverKey? {
        if documentId.hasPrefix("region_") {
            return .region(RegionCode(String(documentId.dropFirst(7))))
        }
        if documentId.hasPrefix("day_"), let date = CalendarDate(iso: String(documentId.dropFirst(4))) {
            return .day(date)
        }
        return nil
    }
}

File: android/data/photo/src/main/kotlin/kr/jjaguk/data/photo/CoversFile.kt
package kr.jjaguk.data.photo

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonObject
import kr.jjaguk.core.model.Cover
import kr.jjaguk.core.model.CoverKey
import kr.jjaguk.core.model.PhotoId

/**
 * `covers.json` 의 모양. `{"region_11140":"a1b2c3", "day_2026-03-05":"d4e5f6"}`
 *
 * 서버에 두든 기기에 두든 **같은 형식**입니다. 혼자 쓰던 짜국을 나중에 같이로 바꿀 때
 * 파일을 그대로 올리면 되도록 하려는 것입니다 (`docs/app/AUTH.md`).
 */
internal object CoversFile {

    private val json = Json { ignoreUnknownKeys = true }

    fun parse(text: String): List<Cover> = try {
        json.parseToJsonElement(text).jsonObject.mapNotNull { (documentId, value) ->
            val photoId = (value as? JsonPrimitive)?.content ?: return@mapNotNull null
            CoverKey.of(documentId)?.let { Cover(it, PhotoId(photoId)) }
        }
    } catch (e: Exception) {
        emptyList()
    }

    fun serialize(covers: List<Cover>): ByteArray = buildString {
        append('{')
        covers.joinTo(this, ",") { "\"" + it.key.documentId + "\":\"" + it.photoId.value + "\"" }
        append('}')
    }.toByteArray()

}
